### ISSUE TITLE
Encapsulate JSON schemas

### ISSUE THREAD

**Issue by delvedor** [fastify/fastify#1351]
## 🚀 Feature Proposal

As briefly discussed in https://github.com/fastify/fastify/issues/1058, we can think about support encapsulated JSON schemas, since currently, we are storing them globally.
If we decide to support this we should land it while we are still in the RC phase of v2, since I would consider it as a breaking change.

## Motivation

Most of our API is encapsulated, so having the schemas encapsulated as well would offer a more complete and coherent experience.

cc @fastify/fastify

**Comment by mcollina** [comment]
I think it would be pretty hard to do in practice, because ajv has its own global registry, which we don’t control.

However, I’m happy to review a PR in that sense. I would try to avoid landing it in v2, as it has been delayed enough.

**Comment by Eomm** [comment]
Hi,
if you agree I would like to work on this task in order to support the json-schema [`definitions`](http://json-schema.org/latest/json-schema-validation.html#rfc.section.9).

I didn't read the code yet, but the main functions to support at the beginning could be:
1. (standard) each route schema can use the `definitions` tag to define reusable json-schema only in that route schema (example below)
1. (standard) the `$ref` tag will reference the json-schema specified on `definitions`
1. (custom for fastify) the shared schema can be referenced in each route schema fields with the pattern `{id}#` (es `my_id#`) as is now, except the `$ref` tag in order to mantain its value standard

Then I don't know if could be usefull to define also a new key `shareddefinitions` in the route schema object that will add a default `definitions` field to each route schema `[body, querystring, params, headers]`


*example:
```js
// definitions are not shared across route schema

const bodyJsonSchema = {
  type: 'array',
  items: { $ref: '#/definitions/positiveInteger' },
  definitions: {
    positiveInteger: {
      type: 'integer',
      exclusiveMinimum: 0
    }
  }
}

const queryStringJsonSchema = {
  type: 'array',
  items: { $ref: '#/definitions/positiveString' },
  definitions: {
    positiveString: {
      type: 'string',
      minLength: 1
    }
  }
}

const schema = {
  body: bodyJsonSchema,
  querystring: queryStringJsonSchema
}
```

**Comment by mcollina** [comment]
Go for it!! I would not add `shareddefinitions` at this point.

**Comment by Eomm** [comment]
This feature should be released in [v2.0.0](https://github.com/fastify/fastify/releases/tag/v2.0.0)

@delvedor can I close it or there is something still missing?

**Comment by delvedor** [comment]
> This feature should be released in v2.0.0

👍
